File: research/primary/reference-ecosystems/ncr-ranger/ranger-app/src/models/scenario.ts

export type Condition = {
  name?: string;
  description?: string;
  command?: string;
  interval?: number;
  source?: Source;
};

export enum ExerciseRole {
  Blue = 'Blue',
  Green = 'Green',
  Red = 'Red',
  White = 'White',
}

export const ExerciseRoleOrder = {
  [ExerciseRole.Blue]: 1,
  [ExerciseRole.Green]: 2,
  [ExerciseRole.Red]: 3,
  [ExerciseRole.White]: 4,
};

export type Entity = {
  name?: string;
  description?: string;
  role?: ExerciseRole;
  mission?: string;
  categories?: string[];
  vulnerabilities?: string[];
  tlos?: string[];
  facts?: Record<string, string>;
  entities?: Record<string, Entity>;
  events?: string[];
};

export type MinScore = {
  absolute?: number;
  percentage?: number;
};

export type Evaluation = {
  name?: string;
  description?: string;
  metrics: string[];
  min_score: MinScore;
};

export type Event = {
  name?: string;
  description?: string;
  source?: Source;
  conditions?: string[];
  injects?: string[];
};

export enum FeatureType {
  Service = 'Service',
  Configuration = 'Configuration',
  Artifact = 'Artifact',
}

export type Feature = {
  name?: string;
  description?: string;
  feature_type: FeatureType;
  source?: Source;
  dependencies?: string[];
  vulnerabilities?: string[];
  variables?: Record<string, string>;
  destination?: string;
};

export type Goal = {
  name?: string;
  description?: string;
  tlos: string[];
};

export type InfraNode = {
  name?: string;
  description?: string;
  count: number;
  links?: string[];
  dependencies?: string[];
};

export type Inject = {
  name?: string;
  description?: string;
  source?: Source;
  from_entity?: string;
  to_entities?: string[];
  tlos?: string[];
};

export enum MetricType {
  Manual = 'Manual',
  Conditional = 'Conditional',
}

export type Metric = {
  name?: string;
  description?: string;
  type: MetricType;
  artifact?: boolean;
  max_score: number;
  condition?: string;
};

export enum NodeType {
  VM = 'VM',
  Switch = 'Switch',
}

export type Resources = {
  ram: number;
  cpu: number;
};

export type Role = {
  username: string;
  entities?: string[];
};

export type Node = {
  type_field: NodeType;
  description?: string;
  resources?: Resources;
  source?: Source;
  features?: Record<string, string>;
  conditions?: Record<string, string>;
  injects?: Record<string, string>;
  vulnerabilities?: string[];
  roles?: Record<string, Role>;
};

export type Script = {
  name?: string;
  description?: string;
  start_time: bigint;
  end_time: bigint;
  speed: number;
  events: string[];
};

export type Source = {
  name: string;
  version: string;
};

export type Story = {
  name?: string;
  clock: bigint;
  scripts: string[];
};

export type TrainingLearningObjective = {
  name?: string;
  description?: string;
  evaluation: string;
};

export type Vulnerability = {
  name: string;
  description: string;
  technical: boolean;
  class: string;
};

export type Scenario = {
  name: string;
  description?: string;
  start: string;
  end: string;
  nodes?: Record<string, Node>;
  features?: Record<string, Feature>;
  infrastructure?: Record<string, InfraNode>;
  conditions?: Record<string, Condition>;
  vulnerabilities?: Record<string, Vulnerability>;
  metrics?: Record<string, Metric>;
  evaluations?: Record<string, Evaluation>;
  tlos?: Record<string, TrainingLearningObjective>;
  entities?: Record<string, Entity>;
  goals?: Record<string, Goal>;
  injects?: Record<string, Inject>;
  events?: Record<string, Event>;
  scripts?: Record<string, Script>;
  stories?: Record<string, Story>;
};

export type TloMapsByRole = {
  [key in ExerciseRole]?: Record<string, TrainingLearningObjective>};

export type ScoringMetadata = {
  startTime: string;
  endTime: string;
  entities: Record<string, Entity>;
  tlos: Record<string, TrainingLearningObjective>;
  evaluations: Record<string, Evaluation>;
  metrics: Record<string, Metric>;
};

File: implementations/python/packages/aces_sdl/scenario.py
"""Top-level Scenario model — the root of the SDL.

The Scenario combines 21 specification sections covering
who (entities, accounts, agents), what (nodes, features,
vulnerabilities, content), when (scripts, stories, events),
and declarative experiment semantics (objectives, scoring
pipeline, conditions, relationships, workflows, variables).

Delivery-level concerns (Docker, Terraform, cloud APIs) are
outside the SDL.
"""

from pydantic import Field, PrivateAttr, model_validator

from ._base import SDLModel
from .accounts import Account
from .agents import Agent
from .conditions import Condition
from .content import Content
from .entities import Entity
from .features import Feature
from .infrastructure import InfraNode
from .nodes import Node
from .objectives import Objective
from .orchestration import Event, Inject, Script, Story, Workflow
from .participant_behavior import ParticipantActionContract, ParticipantObservationBoundary
from .participant_outcome_semantics import OutcomeInterpretationRule
from .relationships import Relationship
from .scoring import TLO, Evaluation, Goal, Metric
from .variables import Variable
from .vulnerabilities import Vulnerability


class ModuleDescriptor(SDLModel):
    """Published module metadata for SDL composition."""

    id: str
    version: str
    parameters: list[str] = Field(default_factory=list)
    exports: dict[str, list[str]] = Field(default_factory=dict)
    description: str = ""

    @model_validator(mode="after")
    def validate_descriptor(self) -> "ModuleDescriptor":
        if "/" not in self.id or self.id.startswith("/") or self.id.endswith("/"):
            raise ValueError("module.id must use canonical 'publisher/name' format")
        if len(self.parameters) != len(set(self.parameters)):
            raise ValueError("module.parameters must be unique")
        for section, names in self.exports.items():
            if len(names) != len(set(names)):
                raise ValueError(f"module.exports.{section} entries must be unique")
        return self


class ImportDecl(SDLModel):
    """A module import expanded before full semantic validation."""

    source: str = ""
    path: str = ""
    namespace: str = ""
    version: str = "*"
    parameters: dict[str, object] = Field(default_factory=dict)
    digest: str = ""

    @model_validator(mode="after")
    def validate_source_fields(self) -> "ImportDecl":
        if not self.source and not self.path:
            raise ValueError("Import requires either 'source' or deprecated 'path'")
        if self.source and self.path:
            raise ValueError("Import may specify only one of 'source' or 'path'")
        if self.path and not self.namespace:
            # Local path imports remain backward compatible and may derive namespace.
            return self
        if self.source.startswith("oci:") and not self.namespace:
            raise ValueError("OCI imports require an explicit namespace")
        return self

    @property
    def normalized_source(self) -> str:
        if self.source:
            return self.source
        return f"local:{self.path}"


class Scenario(SDLModel):
    """Top-level scenario specification.

    A YAML document with up to 21 named sections. Only ``name``
    is required. All sections are optional dicts keyed by
    user-defined identifiers.
    """

    # --- Identity ---
    name: str
    version: str = "*"
    description: str = ""
    module: ModuleDescriptor | None = None
    imports: list[ImportDecl] = Field(default_factory=list)

    # --- OCR SDL: 14 sections ---
    nodes: dict[str, Node] = Field(default_factory=dict)
    infrastructure: dict[str, InfraNode] = Field(default_factory=dict)
    features: dict[str, Feature] = Field(default_factory=dict)
    conditions: dict[str, Condition] = Field(default_factory=dict)
    vulnerabilities: dict[str, Vulnerability] = Field(default_factory=dict)
    metrics: dict[str, Metric] = Field(default_factory=dict)
    evaluations: dict[str, Evaluation] = Field(default_factory=dict)
    tlos: dict[str, TLO] = Field(default_factory=dict)
    goals: dict[str, Goal] = Field(default_factory=dict)
    entities: dict[str, Entity] = Field(default_factory=dict)
    injects: dict[str, Inject] = Field(default_factory=dict)
    events: dict[str, Event] = Field(default_factory=dict)
    scripts: dict[str, Script] = Field(default_factory=dict)
    stories: dict[str, Story] = Field(default_factory=dict)

    # --- Extended sections ---
    content: dict[str, Content] = Field(default_factory=dict)
    accounts: dict[str, Account] = Field(default_factory=dict)
    relationships: dict[str, Relationship] = Field(default_factory=dict)
    agents: dict[str, Agent] = Field(default_factory=dict)
    action_contracts: dict[str, ParticipantActionContract] = Field(default_factory=dict)
    observation_boundaries: dict[str, ParticipantObservationBoundary] = Field(default_factory=dict)
    outcome_interpretation_rules: dict[str, OutcomeInterpretationRule] = Field(default_factory=dict)
    objectives: dict[str, Objective] = Field(default_factory=dict)
    workflows: dict[str, Workflow] = Field(default_factory=dict)
    variables: dict[str, Variable] = Field(default_factory=dict)

    _advisories: list[str] = PrivateAttr(default_factory=list)
    _semantic_validated: bool = PrivateAttr(default=False)
    # Capability-variable provenance carried across the SDL module-import
    # composition boundary. The composition pass strips imported variables
    # from the merged payload by design, so this side-channel preserves the
    # imported variable specs (namespace-prefixed) and the imported nodes'
    # `${name}` refs onto the outer scenario. The instantiation and runtime-
    # compile paths read these to enforce backend `allowed_values` against
    # imported modules even though their variables don't survive the merge.
    # Inner mapping shape mirrors `InstantiatedScenario._node_variable_refs`.
    _module_variable_specs: dict[str, dict[str, object]] = PrivateAttr(default_factory=dict)
    _module_node_variable_refs: dict[str, dict[str, str | None]] = PrivateAttr(default_factory=dict)

    @property
    def advisories(self) -> list[str]:
        """Non-fatal SDL advisories gathered during semantic validation."""
        return list(self._advisories)

    def _set_advisories(self, advisories: list[str]) -> None:
        self._advisories = list(advisories)

    @property
    def semantic_validated(self) -> bool:
        """Whether full semantic validation has already run on this scenario."""
        return self._semantic_validated

    def _set_semantic_validated(self, validated: bool) -> None:
        self._semantic_validated = bool(validated)

    @property
    def module_variable_specs(self) -> dict[str, dict[str, object]]:
        """Namespace-prefixed variable specs preserved across module imports."""
        return {name: dict(spec) for name, spec in self._module_variable_specs.items()}

    @property
    def module_node_variable_refs(self) -> dict[str, dict[str, str | None]]:
        """Refs on imported-module nodes that survived the composition merge."""
        return {name: dict(refs) for name, refs in self._module_node_variable_refs.items()}

    def _set_module_variable_specs(self, specs: dict[str, dict[str, object]]) -> None:
        self._module_variable_specs = {name: dict(spec) for name, spec in specs.items()}

    def _set_module_node_variable_refs(self, refs: dict[str, dict[str, str | None]]) -> None:
        self._module_node_variable_refs = {name: dict(entry) for name, entry in refs.items()}


class InstantiatedScenario(Scenario):
    """Scenario with all `${var}` references resolved to concrete values."""

    _instantiation_parameters: dict[str, object] = PrivateAttr(default_factory=dict)
    _instantiation_profile: str | None = PrivateAttr(default=None)
    # Snapshot of pre-instantiation `${name}` refs on `nodes.os` and
    # `infrastructure.count`, captured by `instantiate_scenario` before
    # substitution. Carries downstream so the runtime processor / planner
    # can reach each variable's `allowed_values` after the resolved values
    # have been written onto the concrete scenario.
    _node_variable_refs: dict[str, dict[str, str | None]] = PrivateAttr(default_factory=dict)

    @property
    def instantiation_parameters(self) -> dict[str, object]:
        """Concrete parameter values used during instantiation."""
        return dict(self._instantiation_parameters)

    @property
    def instantiation_profile(self) -> str | None:
        """Optional instantiation profile name."""
        return self._instantiation_profile

    @property
    def node_variable_refs(self) -> dict[str, dict[str, str | None]]:
        """Pre-instantiation `${var}` refs on `nodes.os` / `infrastructure.count`."""
        return {name: dict(refs) for name, refs in self._node_variable_refs.items()}

    def _set_instantiation_context(
        self,
        *,
        parameters: dict[str, object],
        profile: str | None,
    ) -> None:
        self._instantiation_parameters = dict(parameters)
        self._instantiation_profile = profile

    def _set_node_variable_refs(self, refs: dict[str, dict[str, str | None]]) -> None:
        self._node_variable_refs = {name: dict(entry) for name, entry in refs.items()}


class ExpandedScenario(Scenario):
    """Scenario produced by module/import expansion."""

    _module_namespaces: dict[str, str] = PrivateAttr(default_factory=dict)

    @property
    def module_namespaces(self) -> dict[str, str]:
        return dict(self._module_namespaces)

    def _set_module_namespaces(self, namespaces: dict[str, str]) -> None:
        self._module_namespaces = dict(namespaces)
